### PR TITLE
[forwarder] implement a circuit breaker and use jittered, truncated exponential backoff

### PR DESCRIPTION
**Pull request by ofek**
### What does this PR do?

This implements a circuit breaker and uses a custom exponential backoff algorithm for network error retries. Also, I've enabled all forwarder tests on Windows.

### Resources

- https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter
- https://sre.google/sre-book/addressing-cascading-failures/
- https://docs.microsoft.com/en-us/azure/architecture/patterns/retry
- https://docs.microsoft.com/en-us/azure/architecture/patterns/circuit-breaker

### Motivation

We're forced to overprovision to avoid [cascading failures](https://en.wikipedia.org/wiki/Thundering_herd_problem). It is also just the nice thing to do :smiley:

### Exponential backoff

This [exponential backoff](https://en.wikipedia.org/wiki/Exponential_backoff) algorithm is similar in nature to [Google's HTTP Client Library](https://github.com/google/google-http-java-client), however, unlike that library (and its [Golang port](https://github.com/cenkalti/backoff)), my implementation is thread-safe.

How this works is every network error will increase the time between retries (per endpoint). Each wait time is randomly chosen between exponentially increasing ranges. For example, depending on configuration, `1` error might wait somewhere between `2-4` seconds, `2` errors would wait somewhere between `4-8` seconds, etc. The maximum allowed wait time can be configured.

### Circuit breaker

In order to properly recover from non-transient network failures, I've also implemented a [circuit breaker](https://en.wikipedia.org/wiki/Circuit_breaker_design_pattern). Most research you'll find indicates that the standard way to do this is to send an increasing number of requests until you resume normal execution. This is not the way I do it.

Instead of incrementally increasing the number of allowed requests, I've chosen to exponentially decay the time between each request. So, using the previous example, if we were at `2` errors with a wait of somewhere between `4-8` seconds and we successfully reach the endpoint, we would decrease the error count to `1` error and wait somewhere between `2-4` seconds.

This reduces the complexity of the circuit breaker and also avoids front-loading requests to recovering services.

### Options

- `forwarder_backoff_factor` - This controls the overlap between consecutive retry interval ranges. When set to `2`, there is a guarantee that there will be no overlap. The overlap will asymptotically approach 50% the higher the value is set. Values less then `2` are verboten as there will be range gaps. Default: `2`
- `forwarder_backoff_base` - This controls the rate of exponential growth. Also, you can calculate the start of the very first retry interval range by evaluating the following expression: `forwarder_backoff_base / forwarder_backoff_factor * 2`. Default: `2`
- `forwarder_backoff_max` - This is the maximum number of seconds to wait for a retry. Default: `64`
- `forwarder_recovery_interval` - This controls how many retry interval ranges to step down for an endpoint upon success. Increasing this should only be considered when `forwarder_backoff_max` is particularly high. Default: `1`
- `forwarder_recovery_reset` - Whether or not a successful request should completely clear an endpoint's error count. Default: `false`

### Example times

Here are some examples of expected retry interval ranges (note that the `maxBackoffTime` (currently `64` seconds) will always cap the wait time):

```
minBackoffFactor = 2 // default (no overlap)
baseBackoffTime  = 2 // default
```

| Retries | Range |
| ------- | ----- |
| 1 | [2, 4) |
| 2 | [4, 8) |
| 3 | [8, 16) |
| 4 | [16, 32) |
| 5 | [32, 64) |
| 6 | [64, 128) |
| 7 | [128, 256) |
| 8 | [256, 512) |
| 9 | [512, 1024) |
| 10 | [1024, 2048) |

-----

```
minBackoffFactor = 4 // overlap of 1/3
baseBackoffTime  = 1
```

| Retries | Range |
| ------- | ----- |
| 1 | [0.5, 2) |
| 2 | [1, 4) |
| 3 | [2, 8) |
| 4 | [4, 16) |
| 5 | [8, 32) |
| 6 | [16, 64) |
| 7 | [32, 128) |
| 8 | [64, 256) |
| 9 | [128, 512) |
| 10 | [256, 1024) |

-----

```
minBackoffFactor = 2 // default
baseBackoffTime  = 0.2
```

| Retries | Range |
| ------- | ----- |
| 1 | [0.2, 0.4) |
| 2 | [0.4, 0.8) |
| 3 | [0.8, 1.6) |
| 4 | [1.6, 3.2) |
| 5 | [3.2, 6.4) |
| 6 | [6.4, 12.8) |
| 7 | [12.8, 25.6) |
| 8 | [25.6, 51.2) |
| 9 | [51.2, 102.4) |
| 10 | [102.4, 204.8) |

### Review

@DataDog/croissant @DataDog/metrics-intake

### CODE REVIEW

File: pkg/forwarder/backoff.go
Comment (Viq111) [inline pkg/forwarder/backoff.go]: It's actually random in `[baseBackoffTime / minBackoffFactor * 2; baseBackoffTime * 2[` no ? (when `numErrors == 1`)

File: pkg/forwarder/backoff_test.go
Comment (Viq111) [inline pkg/forwarder/backoff_test.go]: not sure we are testing anything here ;)

File: pkg/forwarder/backoff_test.go
Comment (Viq111) [inline pkg/forwarder/backoff_test.go]: You should actually seed with a known value so your tests are always deterministic and do not depend on when you run them

File: pkg/forwarder/backoff_test.go
Comment (Viq111) [inline pkg/forwarder/backoff_test.go]: Not sure what we really test here, can you add a comment ?

Also a couple of things that's be nice to test:
- We never go above `maxBackoffTime`
- Based on your comments, you know what will `GetBackoffDuration(0)` & `GetBackoffDuration(1)` so we should test those

File: pkg/forwarder/blocked_endpoints.go
Comment (Viq111) [inline pkg/forwarder/blocked_endpoints.go]: I think this should be abstracted into a function in `backoff.go` as this requires someone to know the implementation detail of it

File: pkg/forwarder/blocked_endpoints.go
Comment (Viq111) [inline pkg/forwarder/blocked_endpoints.go]: sometimes instead of doing a lot of casting you can just:
```go
b.nbError += 1
if b.nbError > maxErrors {
  b.nbError = maxErrors
}
```
makes it way more readable ;)

File: pkg/forwarder/blocked_endpoints.go
Comment (Viq111) [inline pkg/forwarder/blocked_endpoints.go]: nit: same here:
```go
b.nbErrors -= recoveryInterval
if b.nbErrors < 0 {
  b.nbErrors = 0
}
```

File: pkg/util/common.go
Comment (Viq111) [inline pkg/util/common.go]: Do those in tests that need them instead

File: pkg/forwarder/backoff_test.go
Comment (Viq111) [reply pkg/forwarder/backoff_test.go]: You can actually do those in tests (mark them as non-parallel if you want) so you really control your seed

**Review by Viq111 — COMMENTED** [review]
Added some nit comments 🙇

File: pkg/forwarder/backoff.go
Comment (ofek) [reply pkg/forwarder/backoff.go]: No actually you can determine the start of the first range using that `baseBackoffTime / minBackoffFactor * 2`

File: pkg/forwarder/backoff_test.go
Comment (ofek) [reply pkg/forwarder/backoff_test.go]: I did this to prevent bad values.

`minBackoffFactor = 1` will have no range e.g.:
```
4 4
8 8
16 16 
32 32 
64 64
```

`1 < minBackoffFactor < 2` will have range gaps e.g. `1.6`:
```
2.5 4 
5 8   
10 16
20 32
40 64
```

File: pkg/forwarder/backoff_test.go
Comment (ofek) [reply pkg/forwarder/backoff_test.go]: We want our results to be random to truly test our algorithm I believe, right?

File: pkg/util/common.go
Comment (ofek) [reply pkg/util/common.go]: The exponential backoff jitter needs this.

File: pkg/forwarder/blocked_endpoints.go
Comment (ofek) [reply pkg/forwarder/blocked_endpoints.go]: I'd really, really prefer not to add overhead to every circuit breaker action when this value is known at compile time and is imo just another implementation detail.

File: pkg/forwarder/blocked_endpoints.go
Comment (ofek) [reply pkg/forwarder/blocked_endpoints.go]: Thanks so much!

File: pkg/forwarder/blocked_endpoints.go
Comment (ofek) [reply pkg/forwarder/blocked_endpoints.go]: Thanks so much!

File: pkg/forwarder/backoff.go
Comment (Viq111) [reply pkg/forwarder/backoff.go]: ah right, didn't see the `start` keyword

File: pkg/forwarder/backoff_test.go
Comment (Viq111) [reply pkg/forwarder/backoff_test.go]: Can you add those comments here then ?
In the future, if we ever change `minBackoffFactor` then people will see the test and edit it too because they wouldn't think of that case

File: pkg/forwarder/backoff_test.go
Comment (Viq111) [reply pkg/forwarder/backoff_test.go]: Here you introduce potential flakiness and not random test.
The problem that will happen here if if this test fails, people will retry once and it will works so they will think there is just flakiness.
If you want to "truly" test the algorithm, you can just have a lot of numbers of tests, i.e: in your test add a `for i := 0; i < 1000; i++ {}` to run your test on 1000 different seeds.

If you even want more than that, you can put it in a benchmark with `for i := 0, i < b.N; i++ {}`
and then run something like `-bench . -benchtime=5m` to thoroughly test.

**Review by Viq111 — DISMISSED** [review]
lgtm from my POV 👍 (I think you'll need a review from croissant too though following repo rules)

File: pkg/forwarder/blocked_endpoints.go
Comment (hush-hush) [inline pkg/forwarder/blocked_endpoints.go]: why not exit, recovering from an endpoint never block should do nothing, no ?

File: pkg/forwarder/forwarder.go
Comment (hush-hush) [inline pkg/forwarder/forwarder.go]: Could we move this logic to the `blockedList` ?

something like ?
```go
if f.blockedList.isClosed(t, retryBefore) {
}
```

File: pkg/forwarder/transaction.go
Comment (hush-hush) [inline pkg/forwarder/transaction.go]: Should we keep `ErrorCount` as a debug tool ? It could be useful to identify one type of transaction causing an endpoint to be closed.

File: pkg/forwarder/backoff.go
Comment (hush-hush) [inline pkg/forwarder/backoff.go]: We should make the back-off behavior configurable since some customers use a proxy (which can have different behavior).

I would add `minBackoffFactor`, `maxBackoffTime`, `baseBackoffTime` and `recoveryInterval` to the configutation. I feel that may be to much. Do you think they are all needed ?

Also should we add on option to completely unblock an endpoint when a transaction is successful. I can imagine some customers asking for it when using a custom proxy.

@ofek what do you think ?

**Review by hush-hush — CHANGES_REQUESTED** [review]
Some small details but great PR overall !

File: pkg/forwarder/blocked_endpoints.go
Comment (ofek) [reply pkg/forwarder/blocked_endpoints.go]: I think it's useful to keep track of endpoints we hit (also great for testing/debugging).

File: pkg/config/config.go
Comment (hush-hush) [inline pkg/config/config.go]: I think we should prefix them with `forwarder_` like we did for the other config option.
Also Could you add some documentation somewhere on how to use those variable. I think the readme of the forwarder would be a good place. I think we should explain what those values do and why we have min/max limit for them.

File: pkg/forwarder/backoff.go
Comment (hush-hush) [inline pkg/forwarder/backoff.go]: I don't think the config is initialized at init since we load it after importing the module. I might be wrong, have you tested it ?

File: pkg/forwarder/backoff.go
Comment (hush-hush) [inline pkg/forwarder/backoff.go]: I think the tests should be updated to use the configuration and also test the min/max limit we impose here.

**Review by hush-hush — CHANGES_REQUESTED** [review]

File: pkg/forwarder/backoff.go
Comment (hush-hush) [inline pkg/forwarder/backoff.go]: I don't think the config is initialized at init since we load it after importing the module. I might be wrong, have you tested it ?

Also, I think the tests should be updated to use the configuration and test the min/max limit we impose here.

**Review by hush-hush — CHANGES_REQUESTED** [review]

File: pkg/forwarder/backoff.go
Comment (ofek) [reply pkg/forwarder/backoff.go]: I did test it. Also, based on https://medium.com/golangspec/init-functions-in-go-eac191b3860a, if `a` imports `b`, any `init` functions in `b` will execute first.

I'll update the tests now.

Comment (hush-hush) [reply pkg/forwarder/backoff.go]: The thing is that the configuration is loaded in the start command of the agent: https://github.com/DataDog/datadog-agent/blob/master/cmd/agent/app/start.go#L125

This is executed after every `init` from every imported module are executed. In Go, `init` functions are executed at import time. I tested it locally (to be sure), and indeed if I set a value in the configuration file it's not taken into account, only the default value is available.

This is also a issue with the testing you added, you don't set any value in the `config` object: https://github.com/DataDog/datadog-agent/pull/1458/commits/5296283c575f646bfcfd746d911d7b85fbc55440#diff-8e4dcd771e37bb4ab78485e4a92a717aR25, you're just testing the default value. We can say that features from viper are working and tested. What we need to test is that the new backoff system do pull value from the configuration and discard erroneous values if any are set in the configuration.

Here an example where we test how a package correctly use the config: https://github.com/DataDog/datadog-agent/blob/master/pkg/util/common_test.go#L139

Comment (ofek) [reply pkg/forwarder/backoff.go]: I understand now, thanks! I've updated this PR.

File: pkg/forwarder/blocked_endpoints.go
Comment (hush-hush) [inline pkg/forwarder/blocked_endpoints.go]: I think we should move the code in `pkg/forwarder/backoff.go` in the `blockerEndpoints` struct. That would avoid having to check if we have already initialized the constant (and get rid of `configLoaded`). Also it remove the init part where we load the default values from the config to overwrite them later and it simplify the code in my opinion.

File: pkg/forwarder/forwarder.go
Comment (hush-hush) [inline pkg/forwarder/forwarder.go]: Why not using `isBlock` since `retryBefore` is basically equal to `time.Now()` https://github.com/DataDog/datadog-agent/blob/308dbf50078d3f64e318951141ad8cdedec5ddfa/pkg/forwarder/forwarder.go#L186

I think we could simply remove `willBeReady`.

**Review by hush-hush — CHANGES_REQUESTED** [review]
Almost good, some nitpick and we should be good to merge

File: pkg/forwarder/blocked_endpoints.go
Comment (ofek) [reply pkg/forwarder/blocked_endpoints.go]: We use the global vars mostly in `GetBackoffDuration`:

https://github.com/DataDog/datadog-agent/pull/1458/files#diff-f5d8ee754e1fc4223fc20e7e6bdd28fdR94

Storing the settings in `blockedEndpoints` instead would require passing an instance of that to the backoff function which would result in too tight of a coupling imo (and would necessitate changing much of this PR).

Also, for testing we also use `loadConfig` heavily to reset default values:

https://github.com/DataDog/datadog-agent/pull/1458/commits/9e01e83b4e78553e11a7a590baf463c21a6d60a8#diff-8e4dcd771e37bb4ab78485e4a92a717a

File: pkg/forwarder/forwarder.go
Comment (ofek) [reply pkg/forwarder/forwarder.go]: Ah I didn't know that, thanks!

File: pkg/forwarder/blocked_endpoints.go
Comment (hush-hush) [reply pkg/forwarder/blocked_endpoints.go]: My point was to make functions from `pkg/forwarder/backoff.go` methods of `blockedEndpoints`. It would simplify the code and testing and change almost nothing in the PR since no functions or constants from `backoff.go` are used anywhere else than inside `blockedEndpoints`.

For testing we would remove `loadConfig` entirely and get the value from the configuration from `newBlockedEndpoints`. This is a pattern we use everywhere else in the agent and is better, in my opinion, than having a constant and global to control initialization.

**Review by hush-hush — APPROVED** [review]

**Pull request merged**
